datafusion/core/src/physical_plan/joins/hash_join.rs

```diff
@@ -377,6 +370,13 @@ impl ExecutionPlan for HashJoinExec {
     ) -> Result<SendableRecordBatchStream> {
         let on_left = self.on.iter().map(|on| on.0.clone()).collect::<Vec<_>>();
         let on_right = self.on.iter().map(|on| on.1.clone()).collect::<Vec<_>>();
+        if self.mode == PartitionMode::Partitioned
+            && self.left.output_partitioning() != self.right.output_partitioning()
+        {
+            return Err(DataFusionError::Plan(format!(
```

Comment:
This is likely caused by some sort of code error, so an internal error may make more sense

```suggestion
            return Err(DataFusionError::Internal(format!(
```

Reply:
The same comment applies to the code below as well